crates/c-api/src/ref.rs

```diff
@@ -43,3 +95,60 @@ pub extern "C" fn wasm_ref_set_host_info_with_finalizer(
     eprintln!("`wasm_ref_set_host_info_with_finalizer` is not implemented");
     std::process::abort();
 }
+
+type wasmtime_externref_finalizer_t = fn(*mut c_void);
+
+struct CExternRef {
+    data: *mut c_void,
+    finalizer: Option<wasmtime_externref_finalizer_t>,
+}
+
+impl Drop for CExternRef {
+    fn drop(&mut self) {
+        if let Some(f) = self.finalizer {
+            f(self.data);
+        }
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn wasmtime_externref_new(data: *mut c_void) -> wasm_val_t {
+    wasm_val_t::from_val(Val::ExternRef(Some(ExternRef::new(CExternRef {
+        data,
+        finalizer: None,
+    }))))
+}
+
+#[no_mangle]
+pub extern "C" fn wasmtime_externref_new_with_finalizer(
+    data: *mut c_void,
+    finalizer: Option<wasmtime_externref_finalizer_t>,
+) -> wasm_val_t {
+    wasm_val_t::from_val(Val::ExternRef(Some(ExternRef::new(CExternRef {
+        data,
+        finalizer,
+    }))))
+}
+
+#[no_mangle]
+pub extern "C" fn wasmtime_externref_data(val: &wasm_val_t, datap: *mut *mut c_void) -> bool {
+    match val.val() {
+        Val::ExternRef(None) => {
+            unsafe {
+                ptr::write(datap, ptr::null_mut());
+            }
+            true
+        }
+        Val::ExternRef(Some(x)) => {
+            let data = match x.data().downcast_ref::<CExternRef>() {
+                Some(r) => r.data,
+                None => x.data() as *const dyn Any as *mut c_void,
```

Comment:
I think this is a bit surprising to do here because this means that you can get back pointers to externref values that you didn't create. I would expect null to be returned here, although I can't imagine this'll be too common of a case to hit since they're no way in the API to create an externref other than the functions above.

Reply:
I was also split on what the best thing to do here was.

I was imagining that someone might have both Rust code and C code that try to communicate via `externref`s, and have some out of band protocol for communicating the types of values they are sending each other.

I can remove it, and return `null` in this case, if you think that is the better course of action.

Reply:
I suppose if there's multiple actors in play you can't really assume anything about the externref pointers you get back anyway. Given two C embeddings there's no guarantee their pointers align, so seems fine to just return what we can here for now. Always possible to tweak later!